app/mailers/mailer.rb

```diff
@@ -77,17 +77,20 @@ def mfa_recommendation_announcement(user_id)
 
   def mfa_required_soon_announcement(user_id)
     @user = User.find(user_id)
-    case @user.mfa_level
-    when "disabled"
-      subject = "[Action Required] Enable multi-factor authentication on your RubyGems account by August 15"
-      @heading = "Enable multi-factor authentication on your RubyGems account"
-    when "ui_only"
-      subject = "[Action Required] Upgrade the multi-factor authentication level on your RubyGems account by August 15"
-      @heading = "Upgrade the multi-factor authentication level on your RubyGems account"
-    end
+    mfa_required_soon_copy = MfaCopy::RequiredSoon.new(@user.mfa_level)
```

Comment:
To resolve a Code Climate violation `similar blocks of code`, the mailers `#mfa_required_soon_announcement` and `#mfa_required_popular_gems_announcement` needed some refactoring. That's why I ended up applying extract class to resolve the violation.

That said, I think the level of repetition in the two mailer actions was acceptable IMO. The mailer actions are small and self-contained. 

I'd argue the indirection brought by this extraction is a worse trade-off.

Reply:
Jenny! I must’ve had the biggest brain fart in the world. Yes, helpers. Duh. This is very much within the responsibilities of a helper. I’ve gone ahead and refactored to use a helper instead.

Inlining the dynamic copy within the string isn’t quite as straight forward as your example. If it was only one variation (per your example), I’d be inclined to make that change too. However, there’s multiple components of dynamic content:
* “Upgrade the” vs. “Enable”
* “Upgrading the” vs. “Enabling”
* “level ” vs. “” (nothing)

It leads to code that’s pretty ugly :see_no_evil:, it’s quite hard to read. I left it here as a [commit](https://github.com/rubygems/rubygems.org/pull/3171/commits/b34b9e4ab9c4ede04a25d962af45858efcccf918) so you can check it out. Ironically, it also gets [flagged](https://user-images.githubusercontent.com/4270287/184578885-acc129b3-9727-40e5-a9aa-f615a7a93f9f.png) by Code Climate too 😅.

---

Re: whether we should guard against the case of the mailer being sent to a user with strong mfa, _within the mailer action itself_, that's a fair question. In my experience, I haven't come across a defensive mailer pattern 🤔. I think the responsibility of ensuring the mailer gets to the right person(s) and conversely, doesn't reach the wrong person(s) should be guarded against and tested at the level in which the mailers get triggered. In this case, at the rake task. 

That said, would you also see us guarding against users with < 180M downloads? I'm wondering if your intention for the guard is to prevent the mailer from ever reaching unintended audiences in general or is it to cover for a missing code path since we're explicitly setting content based on two specific mfa levels. Ultimately, I think the smell this mailer reveals is that it's doing too many jobs 🙈. It could've/should've been separated into two separate mailers with one intended audience each.